src/qibocal/protocols/utils.py

```diff
@@ -757,39 +762,243 @@ def table_html(data: dict) -> str:
     )
 
 
-def extract_feature(
-    x: np.ndarray,
-    y: np.ndarray,
-    z: np.ndarray,
-    find_min: bool,
-):
-    """Extract feature using confidence intervals.
+def euclidean_metric(point1: np.ndarray, point2: np.ndarray):
+    """Euclidean distance between two arrays."""
+    return np.linalg.norm(point1 - point2)
+
+
+def zca_whiten(X):
+    """
+    Applies ZCA whitening to the data (X)
+    https://en.wikipedia.org/wiki/Whitening_transformation
+
+    X: numpy 2d array
+        input data, rows are data points, columns are features
+
+    Returns: ZCA whitened 2d array
+    """
+    assert X.ndim == 2
+    EPS = 10e-5
+
+    #   covariance matrix
+    cov = np.dot(X.T, X)
+    #   d = (lambda1, lambda2, ..., lambdaN)
+    d, E = np.linalg.eigh(cov)
+    #   D = diag(d) ^ (-1/2)
+    D = np.diag(1.0 / np.sqrt(d + EPS))
+    #   W_zca = E * D * E.T
+    W = np.dot(np.dot(E, D), E.T)
+
+    X_white = np.dot(X, W)
+
+    return X_white
+
+
+def filter_data(matrix_z: np.ndarray):
+    """Filter data with a ZCA transformation and then a unit-variance Gaussian."""
+
+    # adding zca filter for filtering out background noise gradient
+    zca_z = zca_whiten(matrix_z)
+    # adding gaussian fliter with unitary variance for blurring the signal and reducing noise
+    return ndimage.gaussian_filter(zca_z, 1)
+
+
+def scaling_global(sig: np.ndarray) -> np.ndarray:
+    """Min–max scaling over the whole np.ndarray (global)."""
+    return scaling_slice(sig, axis=None)
+
+
+def scaling_slice(sig: np.ndarray, axis: Optional[int]) -> np.ndarray:
+    """Min–max scaling over a specific axis of the np.ndarray."""
+    sig_min = np.min(sig, axis=axis)
+    return ((sig.T - sig_min) / (np.max(sig, axis=axis) - sig_min)).T
+
+
+def horizontal_diagonal(xs: np.ndarray, ys: np.ndarray) -> float:
+    """Computing the lenght of the diagonal of a two dimensional grid."""
+    sizes = np.empty(2)
+    for i, values in enumerate([xs, ys]):
+        sizes[i] = np.max(values) - np.min(values)
+    return np.sqrt((sizes**2).sum())
+
+
+def build_clustering_data(peaks_dict: dict, z: np.ndarray):
+    """Preprocessing of the data to cluster."""
+    peaks_x = peaks_dict["x"]["idx"]
+    peaks_y = peaks_dict["y"]["idx"]
+    peaks_sf = z[peaks_y, peaks_x]
```

Comment:
```suggestion
    x_ = peaks_dict["x"]["idx"]
    y_ = peaks_dict["y"]["idx"]
    z_ = z[peaks_y, peaks_x]
```
(so we do not keep the same name for `z`)